### ISSUE TITLE
Migrating to gradle based solution

### ISSUE THREAD

**Issue by yccheok** [Gericop/Android-Support-Preference-V7-Fix#30]
Currently, I have an app which I need to support Android 4 and above.

Previously, I'm using solution from https://github.com/Gericop/Android-Support-Preference-V7-Fix/tree/c42097b2e733aeb2ca69a5c43e9437f6f21ac541

It looks like

![screenshot_20160721-222720](https://cloud.githubusercontent.com/assets/308276/17026861/22141386-4f94-11e6-9bbc-ee5fa95c5051.png)

In previously solution, I'm using multiple styles.xml which looks like

> # values/styles.xml
> 
> ```
> 
> 
> <style name="Theme.JStock.Light" parent="@style/Theme.JStock.Base.Light">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay</item>
>     
>     <item name="android:listSeparatorTextViewStyle">@style/MyListSeparatorTextView</item>
> </style>
> 
> <style name="Theme.JStock.Dark" parent="@style/Theme.JStock.Base.Dark">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay</item>
>     
>     <item name="android:listSeparatorTextViewStyle">@style/MyListSeparatorTextView</item>
> </style>
> 
> 
> 
> 
> 
> <style name="MyListSeparatorTextView">
>     <item name="android:textSize">14sp</item>
>     <item name="android:textStyle">bold</item>
>     <item name="android:paddingTop">16dp</item>
>     <item name="android:layout_marginBottom">16dp</item>
> </style>
> ```
> # values-v14/styles.xml
> 
> <?xml version="1.0" encoding="utf-8"?>
> <resources>
>     
> 
> ```
> 
> 
> <style name="Theme.JStock.Light" parent="@style/Theme.JStock.Base.Light">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay.v14.Material</item>
> </style>
> 
> <style name="Theme.JStock.Dark" parent="@style/Theme.JStock.Base.Dark">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay.v14.Material</item>
> </style>
> 
> 
> <style name="Preference_TextAppearanceMaterialBody2">
>     <item name="android:textSize">14sp</item>
>     <item name="android:fontFamily">sans-serif-medium</item>
>     <item name="android:textStyle">bold</item>
>     <item name="android:textColor">?android:attr/textColorPrimary</item>
> </style>
> 
> 
> ```
> 
> </resources>
> # values-v17/styles.xml
> 
> <?xml version="1.0" encoding="utf-8"?>
> <resources>
>     
> 
> ```
> 
> 
> <style name="Theme.JStock.Light" parent="@style/Theme.JStock.Base.Light">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay.v14.Material</item>
>     
>     <item name="android:listPreferredItemPaddingStart">16dp</item>
>     <item name="android:listPreferredItemPaddingEnd">16dp</item>
> </style>
> 
> <style name="Theme.JStock.Dark" parent="@style/Theme.JStock.Base.Dark">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay.v14.Material</item>
>     
>     <item name="android:listPreferredItemPaddingStart">16dp</item>
>     <item name="android:listPreferredItemPaddingEnd">16dp</item>
> </style>
> 
> 
> <style name="Preference_TextAppearanceMaterialBody2">
>     <item name="android:textSize">14sp</item>
>     <item name="android:fontFamily">sans-serif-medium</item>
>     <item name="android:textStyle">bold</item>
>     <item name="android:textColor">?android:attr/textColorPrimary</item>
> </style>
> 
> 
> ```
> 
> </resources>
> # values-v21/styles.xml
> 
> ```
> 
> 
> <style name="Theme.JStock.Light" parent="@style/Theme.JStock.Base.Light">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay.v14.Material</item>
> </style>
> 
> <style name="Theme.JStock.Dark" parent="@style/Theme.JStock.Base.Dark">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay.v14.Material</item>
> </style>
> ```

After migrating to latest gradle based solution, I was wondering, whether those multiple styles.xml are still necessary? As, I remove all styles.xml except values/styles.xml
# values/styles.xml

> ```
> <style name="Theme.JStock.Light" parent="@style/Theme.JStock.Base.Light">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay</item>
> </style>
> 
> <style name="Theme.JStock.Dark" parent="@style/Theme.JStock.Base.Dark">
>     <item name="preferenceTheme">@style/PreferenceThemeOverlay</item>
> </style>
> ```

But, it looks like this.

![screenshot_20160721-230224](https://cloud.githubusercontent.com/assets/308276/17027727/9fdccdfa-4f97-11e6-9b96-83a48d0036fb.png)

Does that means, I should leave all my multiple styles.xml unchanged?

**Comment by Gericop** [comment]
No, the lib handles the different API versions internally, you only have to use the `@style/PreferenceFixTheme` (or one of its variants) as your parent theme. Also, [**here**](https://github.com/Gericop/Android-Support-Preference-V7-Fix/blob/master/README.md#quick) is a quick getting-started part for other details.

**Comment by yccheok** [comment]
Thanks. I have 2 questions in my minds.

1) Instead of defining `<color name="preference_accent">@color/accent_material_light</color>` in colors.xml, how can I make define this in theme? As, I have 2 different themes for my app, which user can change the theme during run-time. How can I have 2 different `preference_accent`?

2) The left padding doesn't look right. Is there anything I had missed out?

![screenshot_20160722-010324](https://cloud.githubusercontent.com/assets/308276/17032161/7120265c-4faa-11e6-9a5e-c169be0e0b1d.png)

My one and only one `styles.xml` looks like

    <style name="Theme.JStock.Light" parent="@style/Theme.JStock.Base.Light">
        <item name="preferenceTheme">@style/PreferenceFixTheme.Light</item>
    </style>

    <style name="Theme.JStock.Dark" parent="@style/Theme.JStock.Base.Dark">
        <item name="preferenceTheme">@style/PreferenceFixTheme</item>
    </style>

**Comment by Gericop** [comment]
The first is not possible yet unfortunately. If you don't set that, the
default color will be used in the dialogs.

The second is easy to fix: use the PreferenceFixTheme as your main theme's
parent, you don't need to set the overlay manually. Check out the sample
app's styles.xml if you want to see an example.

**Comment by yccheok** [comment]
Hi Gergely,

Thanks for follow up.

Is there possibility to use different preference ascent color under
different app theme? If not, we will just switch back to non grade solution.

As, having to change app theme during runtime is an important feature for
our app.

Thank you.

**Comment by Gericop** [comment]
If it's not a big problem, that the dialogs use the default fallback color (or you choose to use a temporary color that works great for multiple themes), you can continue using this version.

If you really want to change the colors now according to the selected theme, you can add some manual style override to your styles.xml.

# Example 1
If the two themes you use are based on **different variants** of the `PreferenceFixTheme`, for example, one is based on `PreferenceFixTheme`, while the other's parent is `PreferenceFixTheme.Light`, you can override the dialog accent colors simply by adding these to your styles.xml:

```xml
<!-- overriding the normal dialog themes -->
<style name="PreferenceFixTheme.Dialog" parent="@style/Theme.AppCompat.Dialog">
    <item name="colorAccent">#0000ff</item>
</style>

<style name="PreferenceFixTheme.Dialog.Alert" parent="@style/Theme.AppCompat.Dialog.Alert">
    <item name="colorAccent">#0000ff</item>
</style>

<!-- overriding the light variant dialog themes -->
<style name="PreferenceFixTheme.Light.Dialog" parent="@style/Theme.AppCompat.Light.Dialog">
    <item name="colorAccent">#00ff00</item>
</style>

<style name="PreferenceFixTheme.Light.Dialog.Alert" parent="@style/Theme.AppCompat.Light.Dialog.Alert">
    <item name="colorAccent">#00ff00</item>
</style>
```

# Example 2
If you use the **same variant** (e.g. `PreferenceFixTheme.Light.NoActionBar`) as the parent of your themes, the previous solution wouldn't work. In this case, you can specify the dialog styles in the individual themes.

The first theme is called MyTheme01:
```xml
<style name="MyTheme01" parent="@style/PreferenceFixTheme.Light.NoActionBar">
    <!-- other styles -->
    <item name="colorAccent">#00ff00</item>
    <item name="dialogTheme">@style/MyTheme01.Dialog</item>
    <item name="alertDialogTheme">@style/MyTheme01.Dialog.Alert</item>
</style>
<style name="MyTheme01.Dialog" parent="@style/Theme.AppCompat.Light.Dialog">
    <item name="colorAccent">#00ff00</item>
</style>

<style name="MyTheme01.Dialog.Alert" parent="@style/Theme.AppCompat.Light.Dialog.Alert">
    <item name="colorAccent">#00ff00</item>
</style>
```
And your second theme would be MyTheme02:
```xml
<style name="MyTheme02" parent="@style/PreferenceFixTheme.Light.NoActionBar">
    <!-- other styles -->
    <item name="colorAccent">#0000ff</item>
    <item name="dialogTheme">@style/MyTheme02.Dialog</item>
    <item name="alertDialogTheme">@style/MyTheme02.Dialog.Alert</item>
</style>
<style name="MyTheme02.Dialog" parent="@style/Theme.AppCompat.Light.Dialog">
    <item name="colorAccent">#0000ff</item>
</style>

<style name="MyTheme02.Dialog.Alert" parent="@style/Theme.AppCompat.Light.Dialog.Alert">
    <item name="colorAccent">#0000ff</item>
</style>
```

Hope this helps.

---
In the future I plan to overcome this problem by resolving the accent color for the dialogs programmatically, but it's not necessarily an easy job to do.

**Comment by yccheok** [comment]
By using `DIVIDER_CATEGORY_BEFORE_FIRST` as follow

    public class JStockPreferenceFragment extends com.takisoft.fix.support.v7.preference.PreferenceFragmentCompatDividers implements SharedPreferences.OnSharedPreferenceChangeListener, PreferenceFragmentCompat.OnPreferenceDisplayDialogCallback {

        @Override
        public void onActivityCreated(Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);

            this.setDividerPreferences(DIVIDER_CATEGORY_BEFORE_FIRST);
        }

This is the closest divider effect I can get

![screenshot_20160725-020026](https://cloud.githubusercontent.com/assets/308276/17085492/f9ffd27a-520b-11e6-8ae8-8fc7a5c12805.png)

If you look carefully, there's a divider line, between toolbar shadow and "UI" 1st category.

I would like to skip divider from drawn on 1st category.

In my previous `PreferenceFragmentCompatFix`, I had done some modification, to prevent divider from drawn in 1st category.

    private void createDividers() {
        if (itemDecoration != null)
            return;

        itemDecoration = new RecyclerView.ItemDecoration() {
            @Override
            public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
                LinearLayoutManager lm = (LinearLayoutManager) parent.getLayoutManager();

                final int first = lm.findFirstVisibleItemPosition();
                final int last = lm.findLastVisibleItemPosition();

                RecyclerView.Adapter adapter = parent.getAdapter();

                for (int i = first; i <= last; i++) {
                    // Skip the first item.
                    if (i == 0) {
                        continue;
                    }

For this gradle based solution, is there any way I can achieve similar objective?

**Comment by Gericop** [comment]
You can use the `DIVIDER_PREFERENCE_AFTER_LAST` (it's also mentioned in the method's description), this will do the same except the dividers appearing after the last preferences in the preference groups.

**Comment by yccheok** [comment]
Thanks. Thing works as expected. After trying, I really love your gradle solution. It makes the maintenance work much more easier.

![android4-dark](https://cloud.githubusercontent.com/assets/308276/17104404/58c5900c-52b5-11e6-8565-19fa6930a652.png)
![android4-light](https://cloud.githubusercontent.com/assets/308276/17104402/58c236e6-52b5-11e6-9a34-f7b096e747f2.png)
![android6-dark](https://cloud.githubusercontent.com/assets/308276/17104403/58c268aa-52b5-11e6-88a9-63cb9c1b457a.png)
![android6-light](https://cloud.githubusercontent.com/assets/308276/17104405/58caaf10-52b5-11e6-98a9-b3fb0ecb40cd.png)

**Comment by Gericop** [comment]
Sure, no problem. I'll try to fix that fallback color thing in order to eliminate the extra work needed for overriding dialog accent colors.

**Comment by yccheok** [comment]
May I know is there any Proguard settings need to take into consideration, as reflection method is being used in https://github.com/Gericop/Android-Support-Preference-V7-Fix/commit/7de016b007e28a264001a8bb353f110a7f64bb69

Thanks.

**Comment by Gericop** [comment]
If your ProGuard settings leave the support things alone, it will be fine.

If you, however, let ProGuard obfuscate those things, it might get messy. There are currently 3 classes that use reflection:

- `PreferenceManagerFix`: this shouldn't be a problem because the field is identified by its class type
- `PreferenceFragmentCompat`: the field is also identified by its class type
- `PreferenceGroupAdapter`: this could be a problem, because all of the reflections are made using the field names rather than the class types

In order to overcome this, you have to make sure that the ProGuard won't obfuscate the following field and class names:

- `android.support.v7.preference.PreferenceGroupAdapter.`**`mPreferenceLayouts`**
- `android.support.v7.preference.PreferenceGroupAdapter.`**`PreferenceLayout`** (this is an inner class)
- `android.support.v7.preference.PreferenceGroupAdapter.PreferenceLayout.`**`resId`**
- `android.support.v7.preference.PreferenceGroupAdapter.PreferenceLayout.`**`widgetResId`**

I hope I could help.

**Comment by yccheok** [comment]
@Gericop 

Before running proguard, everything works fine including API 16.

I realize padding information is gone in API 16, after I export the app to APK (running proguard).

I added the following in proguard.

    # https://github.com/Gericop/Android-Support-Preference-V7-Fix/issues/30
    -keep class android.support.v7.preference.PreferenceGroupAdapter {
    *;
    }
    -keep class com.takisoft.fix.support.v7.preference.PreferenceGroupAdapter {
    *;
    }

Strangely. The padding information still gone. I check on the error log message. I expect to see some exception being thrown. However, there's no exception related to reflection.

Any idea?

**Comment by yccheok** [comment]
I think I had found out the solution. The following works for me.

    # https://github.com/Gericop/Android-Support-Preference-V7-Fix/issues/30
    -keep class android.support.v7.preference.PreferenceGroupAdapter {
    *;
    }
    -keep class android.support.v7.preference.PreferenceGroupAdapter$PreferenceLayout {
    *;
    }
    -keepattributes SourceFile,LineNumberTable,InnerClasses

**Comment by Gericop** [comment]
The official support lib v24.2.0 fixes the preference accent color fallback thing, thus it's not required (nor possible) to override `preference_accent` since v24.2.0.0. From now on, it's possible to just set the `colorAccent` attribute of a style which enables usage of multiple, different-colored themes.

**Comment by Gericop** [comment]
v24.2.0.0 has been released, update your gradle file:
```gradle
compile 'com.takisoft.fix:preference-v7:24.2.0.0'
```

**Comment by Gericop** [comment]
**v25.1.0.2** solves the ProGuard problem too.
